src/cobra/core/gene.py

```diff
@@ -299,3 +294,289 @@ def _repr_html_(self):
             n_reactions=len(self.reactions),
             reactions=format_long_string(", ".join(r.id for r in self.reactions), 200),
         )
+
+
+class GPR(Module):
+    """A Gene Reaction rule in a cobra model, using AST as base class
+
+    Parameters
+    ----------
+    string_gpr : string
+        A GPR in string format, which will be parsed into AST
+    """
+
+    def __init__(self, string_gpr=None, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._genes = set()
+        if string_gpr:
+            self.from_string(string_gpr)
+
+    def from_string(self, string_gpr) -> None:
+        """
+
+        Parameters
+        ----------
+        string_gpr: str
+            a string that describes the gene rules, in a format like
+            A & B
+
+        Returns
+        -------
+        Does not return
+
+            Sets self.body as
+            Parsed AST tree that has the gene rules
+            This function also sets self._genes with the gene ids in the AST
+
+        """
+        uppercase_AND = re.compile(r"\bAND\b")
+        uppercase_OR = re.compile(r"\bOR\b")
+        str_expr = string_gpr.strip()
+        if len(str_expr) == 0:
+            self.body = []
+            return
+        for char, escaped in replacements:
+            if char in str_expr:
+                str_expr = str_expr.replace(char, escaped)
+        escaped_str = keyword_re.sub("__cobra_escape__", str_expr)
+        escaped_str = number_start_re.sub("__cobra_escape__", escaped_str)
+
+        try:
+            tree = ast_parse(escaped_str, "<string>", "eval")
+        except (SyntaxError, TypeError) as e:
+            if "AND" in string_gpr or "OR" in string_gpr:
+                warn(
+                    "uppercase AND/OR found in rule '%s' for '%s'"
+                    % (string_gpr, repr(self))
+                )
+                string_gpr = uppercase_AND.sub("and", string_gpr)
+                string_gpr = uppercase_OR.sub("or", string_gpr)
+            try:
+                tree = ast_parse(string_gpr, "<string>", "eval")
+            except SyntaxError as e:
+                warn(
+                    "malformed gene_reaction_rule '%s' for %s"
+                    % (string_gpr, repr(self))
+                )
+                warn("GPR will be empty")
+                warn(e.msg)
+                return None
+        cleaner = GPRCleaner()
+        cleaner.visit(tree)
+        self._genes = cleaner.gene_set
+        if "" in self._genes:
+            self._genes.remove("")
+        self.body = tree.body
+        self.eval()  # ensure the rule can be evaluated
+
+    @property
+    def geneset(self):
+        self.updategenes()
+        return frozenset(self._genes)
+
+    def updategenes(self):
```

Comment:
Try to separate individual words in function names, so better call it `update_genes`.